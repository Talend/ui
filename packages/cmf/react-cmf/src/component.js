import invariant from 'invariant';
import actionCreator from './actionCreator';
import expression from './expression';
import sagas from './sagas';
import registry from './registry';
import CONST from './constant';

/**
 * All stuff related to the routing in CMF
 * @module react-cmf/lib/component
 * @see module:react-cmf/lib/component
 */

/**
 * return a component from the registry
 * @param  {string} id the component id you want
 * @param  {object} context optional context to get the registry
 * @return {function} the react component
 */
function get(id, context) {
	const component = registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context);
	if (!component) {
		throw new Error(`component not found in the registry: ${id}`);
	}
	return component;
}

/**
 * register a component to let CMF be able to Inject it.
 * @param  {string} id the component id you want to register
 * @param  {any} component the component you want to register
 * @param  {object} context optional context to get the registry
 */
function register(id, component, context) {
	if (!component) {
		invariant(
			process.env.NODE_ENV === 'production',
			'You cannot register undefined as a component for id "%s"',
			id,
		);
		return;
	}
	registry.addToRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, component, context);
	if (component.actions) {
		actionCreator.registerMany(component.actions, context);
	}
	if (component.expressions) {
		expression.registerMany(component.expressions, context);
	}
	if (component.sagas) {
		sagas.registerMany(component.sagas, context);
	}
}

const registerMany = registry.getRegisterMany(register);

function has(id, context) {
	return (
		registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context) !== undefined
	);
}

export default {
	get,
	has,
	register,
	registerMany,
};
