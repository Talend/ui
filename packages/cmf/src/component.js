import invariant from 'invariant';
import actionCreator from './actionCreator';
import expression from './expression';
import registry from './registry';
import CONST from './constant';

/**
 * All stuff related to the routing in CMF
 * @module react-cmf/lib/component
 * @see module:react-cmf/lib/component
 */

/**
 * return a component from the registry
 * @param  {string} id the component id you want
 * @param  {object} context optional context to get the registry
 * @return {function} the react component
 */
function get(id, context) {
	const component = registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context);
	if (!component) {
		throw new Error(`component not found in the registry: ${id}`);
	}
	return component;
}

/**
 * register a component for the router configuration
 * @param  {string} id the component id you want to register
 * @param  {any} component the component you want to register
 * @param  {object} context optional context to get the registry
 */
function register(id, component, context) {
	if (!component) {
		invariant(process.env.NODE_ENV !== 'production', 'You can register undefined as a component');
		return;
	}
	registry.addToRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, component, context);
	if (component.actions) {
		Object.keys(component.actions).forEach(key => {
			actionCreator.register(key, component.actions[key], context);
		});
	}
	if (component.expressions) {
		Object.keys(component.expressions).forEach(key => {
			expression.register(key, component.expressions[key], context);
		});
	}
}

const registerMany = registry.getRegisterMany(register);

function has(id, context) {
	return (
		registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context) !== undefined
	);
}

export default {
	get,
	has,
	register,
	registerMany,
};
