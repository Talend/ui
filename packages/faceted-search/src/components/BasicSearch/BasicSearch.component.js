import React, { useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { getTheme } from '@talend/react-components/lib/theme';
import get from 'lodash/get';

import { AddFacetPopover } from '../AddFacetPopover';
import { BadgeOverlay } from '../Badges';
import { BadgesGenerator } from '../BadgesGenerator';
import { generateBadge } from '../types/badgeDefinition.type';
import { useFacetedSearchContext } from '../context/facetedSearch.context';
import { BadgeFacetedProvider } from '../context/badgeFaceted.context';

import { createBadgesDict, getBadgesFromDict } from '../../dictionary/badge.dictionary';
import { createOperatorsDict, getOperatorsFromDict } from '../../dictionary/operator.dictionary';
import { useFacetedBadges, BADGES_ACTIONS } from '../../hooks/facetedBadges.hook';
import { badgesFacetedPropTypes, operatorsPropTypes } from '../facetedSearch.propTypes';

import theme from './BasicSearch.scss';

const css = getTheme(theme);

const isInCreation = badge => get(badge, 'metadata.isInCreation', true);

const BasicSearch = ({
	badgesDefinitions,
	badgesFaceted,
	customBadgesDictionary,
	customOperatorsDictionary,
	initialFilterValue,
	onSubmit,
	setBadgesFaceted,
}) => {
	const { id, t } = useFacetedSearchContext();
	const operatorsDictionary = useMemo(() => createOperatorsDict(t, customOperatorsDictionary));
	const badgesDictionary = useMemo(() => createBadgesDict(customBadgesDictionary));
	const [state, dispatch] = useFacetedBadges(badgesFaceted, setBadgesFaceted);

	useEffect(() => {
		if (!state.badges.some(isInCreation)) {
			onSubmit({}, state.badges);
		}
	}, [state.badges, onSubmit]);

	const onClickOverlayRow = setOverlayOpened => (_, badgeDefinition) => {
		const operators = getOperatorsFromDict(
			operatorsDictionary,
			get(badgeDefinition, 'metadata.operators'),
		);
		dispatch(BADGES_ACTIONS.add(generateBadge(operators)(badgeDefinition)));
		setOverlayOpened(false);
	};
	const basicSearchId = `${id}-basic-search`;
	const badgeFacetedContextValue = { state, dispatch, onSubmit };
	return (
		<div id={basicSearchId} className={css('tc-basic-search')}>
			<BadgeFacetedProvider value={badgeFacetedContextValue}>
				<BadgesGenerator
					badges={state.badges}
					badgesDictionary={badgesDictionary}
					getBadgeFromDict={getBadgesFromDict}
					id={basicSearchId}
					t={t}
				/>
			</BadgeFacetedProvider>
			<BadgeOverlay
				id={basicSearchId}
				iconName="plus-circle"
				label={t('OPEN_ADD_FACET_BUTTON', { defaultValue: 'Add filter' })}
				hideLabel
			>
				{setOverlayOpened => (
					<AddFacetPopover
						badgesDefinitions={badgesDefinitions}
						id={basicSearchId}
						initialFilterValue={initialFilterValue}
						onClick={onClickOverlayRow(setOverlayOpened)}
						t={t}
					/>
				)}
			</BadgeOverlay>
		</div>
	);
};

BasicSearch.propTypes = {
	badgesDefinitions: badgesFacetedPropTypes,
	badgesFaceted: PropTypes.shape({
		badges: badgesFacetedPropTypes,
	}),
	customBadgesDictionary: PropTypes.object,
	customOperatorsDictionary: operatorsPropTypes,
	initialFilterValue: PropTypes.string,
	onSubmit: PropTypes.func.isRequired,
	setBadgesFaceted: PropTypes.func,
};

// eslint-disable-next-line import/prefer-default-export
export { BasicSearch };
