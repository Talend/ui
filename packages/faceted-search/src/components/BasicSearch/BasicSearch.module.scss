@use '~@talend/design-tokens/lib/tokens';

.tc-basic-search {
	align-items: center;
	display: flex;
	height: inherit;
	overflow: visible;
	position: relative;
	width: 100%;
	padding-left: tokens.$coral-spacing-s;
	padding-right: tokens.$coral-spacing-s;

	&-quicksearch {
		margin-right: tokens.$coral-spacing-xs;
	}

	&-content {
		display: flex;
		align-items: center;
		flex-direction: row;
		flex-wrap: wrap;
		flex-grow: 1;
	}
}
