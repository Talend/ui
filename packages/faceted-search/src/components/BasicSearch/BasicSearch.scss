.tc-basic-search {
	align-items: center;
	display: flex;
	flex-wrap: wrap;
	height: inherit;
	margin-right: $padding-large;
	overflow: auto;
	padding-left: $padding-large;
	width: calc(100% - 15rem);
}
