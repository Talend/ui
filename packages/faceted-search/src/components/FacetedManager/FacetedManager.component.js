import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { I18N_DOMAIN_FACETED_SEARCH } from '../../constants';
import { FacetedSearchProvider } from '../context/facetedSearch.context';

const FacetedManager = ({ children, id, inProgress, error }) => {
	const { t } = useTranslation(I18N_DOMAIN_FACETED_SEARCH);
	const contextValue = {
		error,
		id,
		inProgress,
		t,
	};
	return <FacetedSearchProvider value={contextValue}>{children}</FacetedSearchProvider>;
};

FacetedManager.propTypes = {
	children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
	id: PropTypes.string.isRequired,
	inProgress: PropTypes.bool,
	error: PropTypes.string,
};

// eslint-disable-next-line import/prefer-default-export
export { FacetedManager };
