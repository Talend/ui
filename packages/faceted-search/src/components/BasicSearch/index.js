import { BasicSearch } from './BasicSearch.component';

export { BasicSearch };
