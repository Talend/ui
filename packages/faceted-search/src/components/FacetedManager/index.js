import { FacetedManager } from './FacetedManager.component';

export { FacetedManager };
