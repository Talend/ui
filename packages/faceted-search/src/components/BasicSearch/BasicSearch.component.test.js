/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';

import { BasicSearch } from './BasicSearch.component';
import { FacetedManager } from '../FacetedManager';

describe('BasicSearch', () => {
	const badgeText = {
		properties: {
			attribute: 'name',
			initialOperatorOpened: false,
			initialValueOpened: false,
			label: 'Name',
			operator: { label: 'Equals', name: '=', iconName: 'equal' },
			operators: [
				{ label: 'Equals', name: '=', iconName: 'equal' },
				{ label: 'Contains', name: 'contains', iconName: 'contains' },
			],
			type: 'text',
			value: 'hello',
		},
		metadata: {
			badgeId: 'name-7bc9bd07-3b46-4b8c-a406-a08b6263de5b',
			badgePerFacet: 'N',
			isInCreation: true,
			entitiesPerBadge: '1',
			operators: ['contains', '='],
		},
	};
	const badgesFaceted = {
		badges: [{ ...badgeText }],
	};
	const badgeDefinitionName = {
		properties: {
			attribute: 'name',
			initialOperatorOpened: true,
			initialValueOpened: false,
			label: 'Name',
			operator: {},
			operators: [],
			type: 'text',
		},
		metadata: {
			badgePerFacet: 'N',
			entitiesPerBadge: '1',
			operators: ['contains', '='],
		},
	};

	const badgesDefinitions = [badgeDefinitionName];

	it('should render the default html output with no badges', () => {
		// Given
		const props = {
			onSubmit: jest.fn(),
		};
		// When
		const wrapper = mount(
			<FacetedManager id="manager-id">
				<BasicSearch {...props} />
			</FacetedManager>,
		);
		// Then
		expect(wrapper.html()).toMatchSnapshot();
	});
	it('should render the default html output with some badges', () => {
		// Given
		const props = {
			badgesDefinitions,
			badgesFaceted,
			onSubmit: jest.fn(),
		};
		// When
		const wrapper = mount(
			<FacetedManager id="manager-id">
				<BasicSearch {...props} />
			</FacetedManager>,
		);
		// Then
		expect(wrapper.html()).toMatchSnapshot();
	});
	it('should not trigger onSubmit when a badge is in creation', () => {
		// given
		const onSubmit = jest.fn();
		const props = {
			badgesDefinitions,
			badgesFaceted,
			onSubmit,
		};
		// when
		const wrapper = mount(
			<FacetedManager id="manager-id">
				<BasicSearch {...props} />
			</FacetedManager>,
		);
		wrapper.update();
		// then
		expect(onSubmit).not.toHaveBeenCalled();
	});
	it('should trigger onSubmit when no badge is in creation', () => {
		// given
		const onSubmit = jest.fn();
		const props = {
			badgesDefinitions,
			badgesFaceted: {
				badges: [
					{
						...badgeText,
						metadata: {
							...badgeText.metadata,
							isInCreation: false,
						},
					},
				],
			},
			onSubmit,
		};
		// when
		const wrapper = mount(
			<FacetedManager id="manager-id">
				<BasicSearch {...props} />
			</FacetedManager>,
		);
		wrapper.update();
		// then
		expect(onSubmit).toHaveBeenCalled();
		expect(onSubmit.mock.calls.length).toBe(1);
		expect(onSubmit.mock.calls[0][0]).toEqual({});
		expect(onSubmit.mock.calls[0][1]).toEqual(props.badgesFaceted.badges);
	});
});
