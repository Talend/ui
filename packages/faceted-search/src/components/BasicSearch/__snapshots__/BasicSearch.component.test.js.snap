// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicSearch should render the default html output with no badges 1`] = `
<div id="manager-id-basic-search"
     class="tc-basic-search theme-tc-basic-search"
>
  <div>
    <button id="manager-id-basic-search-action-overlay"
            aria-label="Add facets"
            type="button"
            class="btn btn-link"
    >
      <svg name="talend-plus-circle"
           class="theme-tc-svg-icon tc-svg-icon"
           focusable="false"
           aria-hidden="true"
      >
        <use xlink:href="#talend-plus-circle">
        </use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`BasicSearch should render the default html output with some badges 1`] = `
<div id="manager-id-basic-search"
     class="tc-basic-search theme-tc-basic-search"
>
  <div>
    <button id="manager-id-basic-search-action-overlay"
            aria-label="Add facets"
            type="button"
            class="btn btn-link"
    >
      <svg name="talend-plus-circle"
           class="theme-tc-svg-icon tc-svg-icon"
           focusable="false"
           aria-hidden="true"
      >
        <use xlink:href="#talend-plus-circle">
        </use>
      </svg>
    </button>
  </div>
</div>
`;
